### PR TITLE
Fix request flow and fingerprint

### PR DESCRIPTION
**Pull request by bilfeldt**
This PR implements modifications to the middleware and now does the tracking pre-controller (issue #40) while it also introduces a new request macro `$request->footprint()` which uses a new `Fingerprinter` class.

Missing:
- ~Update `README.md` as the interfaces have changed~ (was not needed)
- [x] Find a way to make the `footprint()` macro configurable (perhaps a new `Footprinting` class which implements a `FootprintingInterface`?)

@kyranb This is inspired from you suggestion on a fingerprint method. I think for this to be really useful then it should be possible to change the logic used for generating the footprint. Do you have any suggestions how we can do that? Making this configurable would be a good way for SPAs to do some customisation like providing the footprint as a query parameter from the SPA.

### CODE REVIEW

**Comment by bilfeldt** [comment]
@kyranb I went ahead a implemented a custom `Footprinter` class which is responsible for generating the string that is saved in the cookie. By disabling the configuration `uniqueness` then this string will not be random but instead created from the ip address and the `User-agent` header.

Try to set the configuration `'uniqueness' => false` and do the following:
- Visit any tracked route: `fingerprint1example` will be saved
- Change your IP (use your mobiles net for example): `fingerprint1example` will still be saved since it uses cookies.
- Visit a route from a Chrome Incognito: `fingerprint2example` will be generated since it does not have an existing cookie to use
- Change the IP back (use your wifi again) and open a new incognito window: `fingerprint1example` will be saved although it does not use the cookie but because the ip and the `user-agent` is the same - this is cool in my opinion! 👍 
 - Visit another route from a Safari Incognito: `fingerprint3example` will be generated since it does not have access to the cookie set by chrome and the `user-agent` is different
 
 I think this is a pretty neat feature - thanks for suggesting that :) Let me know what you think @kyranb .

File: README.md [hunk 1]
```diff
@@ -183,6 +186,17 @@ $user = User::find(1);
 $user->finalAttributionData();
 ```
 
+##### Trackng process in details
+First off the `CaptureAttributionDataMiddleware` can be registred globally or on a selected list of routes.
```
Comment (kyranb) [inline README.md:199]: Just a small typo here:)

registred > registered

File: README.md [hunk 2]
```diff
@@ -183,6 +186,17 @@ $user = User::find(1);
...
+
+Whenever an incomming request passes through the `CaptureAttributionDataMiddleware` middleware then it checks whether or not the request should be tracked using the class `TrackingFilter` (can be changed to any class implementing the `TrackingFilterInterface`) and if the request should be logged `TrackingLogger` will do so (can be changed to any class implementing `TrackingLoggerInterface`).
+
+The `TrackingLogger` is responsible for logging relevant information about the request as a `Vist` record. The most important parameter is the requests "footprint" which is the entity that *should* be the same for multiple requests performed by the same user and hence this is what is used to link different requests.
```
Comment (kyranb) [inline README.md:203]: requests > request's

**Review by kyranb — COMMENTED** [review]
Thanks so much for taking this package to a whole new level :)

**Comment by bilfeldt** [comment]
> Thanks so much for taking this package to a whole new level :)

It is a pleasure @kyranb  👍 Typos fixed now.

Since v3 will be a breaking change, wouldn't it be a good time to change support to say Laravel 7+ and PHP 7.4+? That would also allow for return type declarations and typed class properties.

Another thought is whether or not to rely on the awesome work of the people at Spatie and rely on this package: https://github.com/spatie/laravel-http-logger/tree/master/src
it would make sense instead of reinventing the wheel, but on the other hand, then that package will dictate the supported PHP and Laravel versions.

**Comment by kyranb** [comment]
You're on an absolute roll, thanks so much for your contributions Bilfeldt :) Glad to hear that this package I made a few years ago is still getting used. I've often wondered how many $$$ it's saved from being wasted ha!

I agree, on `master` `composer.json` is set to require `"laravel/framework": "^8.0",`. So yes, before tagging v3 we will need to clarify that is a new requirement.

Happy to accept a PR for type declarations and typed class properties too 👍

**Comment by kyranb** [comment]
Mmm I think tying it the Spatie package might end up complicating things as this package's core functionality seems to work as is at the moment?

**Comment by bilfeldt** [comment]
> Mmm I think tying it the Spatie package might end up complicating things as this package's core functionality seems to work as is at the moment?

I have no strong opinion regarding this seeing as it is such a small and simple part. Just wanted the share the possibility.

**Pull request merged**